src/spaces/compartment.jl

```diff
@@ -0,0 +1,188 @@
+  Iterators.filter(!isequal(pos), nn)
+end
+
+function positions(model::ABM{<:AbstractAgent,<:CompartmentSpace})
+  x = CartesianIndices(model.space.grid.s)
+  return (Tuple(y) for y in x)
+end
+
+function ids_in_position(pos::ValidPos, model::ABM{<:AbstractAgent,<:CompartmentSpace})
+  return model.space.grid.s[pos...]
+end
+
+cell_center(pos::Tuple) = getindex.(modf.(pos), 2) .+ 0.5
+distance_from_cell_center(pos::Tuple, center) = sqrt(sum(abs2.(pos .- center)))
+
+"""
+    space_neighbors(position, model::ABM, r=1; kwargs...) → ids
+
+Return an iterator of the ids of the agents within "radius" `r` of the given `position`
+(which must match type with the spatial structure of the `model`).
+
+What the "radius" means depends on the space type:
+- `GraphSpace`: `r` means the degree of neighbors in the graph and is an integer.
+  For example, for `r=2` include first and second degree neighbors.
+- `GridSpace, ContinuousSpace`: Standard distance implementation according to the
+  underlying space metric.
+
+## Keywords
+Keyword arguments are space-specific.
+For `GraphSpace` the keyword `neighbor_type=:default` can be used to select differing
+neighbors depending on the underlying graph directionality type.
+- `:default` returns neighbors of a vertex. If graph is directed, this is equivalent
+  to `:out`. For undirected graphs, all options are equivalent to `:out`.
+- `:all` returns both `:in` and `:out` neighbors.
+- `:in` returns incoming vertex neighbors.
+- `:out` returns outgoing vertex neighbors.
+"""
+function space_neighbors(pos, model, r=1; exact=false)
+  center = cell_center(pos)
+  focal_cell = ceil.(Int, center)
+  if exact
+    newr = ceil.(Int, r)
+    corner_of_largest_square_in_circle = floor.(Int, center .+ newr)
+    max_distance = corner_of_largest_square_in_circle[1] - focal_cell[1]
+    final_ids = Int[] # TODO make it an iterator
+    allcells = nearby_positions(focal_cell, model, newr)
+    for cell in allcells
+      if !any(x-> x> max_distance, abs.(cell .- focal_cell)) # certain cell
+        ids = ids_in_position(cell, model)
+        final_ids = vcat(final_ids, ids)
+      else # uncertain cell
+        ids = ids_in_position(cell, model)
+        filter!(i -> sqrt(sum(abs2.(center .- model[i]))) ≤ rnew, ids)
+        final_ids = vcat(final_ids, ids)
+      end
+    end
+    return final_ids
+  else
+    δ = distance_from_cell_center(pos, center)
+    newr = ceil.(Int, r+δ)
+    return nearby_ids(focal_cell, model, newr)
+  end
+end
+
+"""
+    node_neighbors(position, model::ABM, r=1; kwargs...) → positions
+
+Return an iterator of all positions within "radius" `r` of the given `position`
+(which excludes given `position`).
+The `position` must match type with the spatial structure of the `model`).
+
+The value of `r` and possible keywords operate identically to [`space_neighbors`](@ref).
+"""
+node_neighbors(position, model, r=1; exact=false) = notimplemented(model)
```

Comment:
This shouldn't be here at all. Why duplicate it? the existing implementation in the API file already does this.